### ISSUE TITLE
Massively slows down startup

### ISSUE THREAD

**Issue by dccsillag** [dccsillag/magma-nvim#6]
When my computer is under a moderate load, this plugin appears to be massively slowing down opening time: 1 more second!

Here's the start of the output of [`:StartupTime`](https://github.com/dstein64/vim-startuptime):

```
       startup: 1949.6
event                  time percent plot
BufEnter autocommand 954.79   48.97 ██████████████████████████
packer_compiled.lua  194.39    9.97 █████▎
expanding arguments  155.69    7.99 ████▎
python3.vim          149.25    7.66 ████▏
init.lua             112.73    5.78 ███▏
csillag.vim           56.51    2.90 █▌
filetype.vim          44.51    2.28 █▎
loading packages      43.17    2.21 █▏
opening buffers       39.72    2.04 █▏
loading plugins       32.92    1.69 ▉
syntax.vim            18.02    0.92 ▌
synload.vim           17.45    0.90 ▌
csillag.vim           16.87    0.87 ▌
rplugin.vim           14.79    0.76 ▍
rplugin.vim           14.11    0.72 ▍
indent_blankline.vim  13.74    0.70 ▍
reading ShaDa         12.14    0.62 ▍
tcomment.vim          11.61    0.60 ▍
TabularMaps.vim        9.48    0.49 ▎
wordmotion.vim         8.41    0.43 ▎
entire.vim             8.13    0.42 ▎
first screen update    7.63    0.39 ▎
```

**Comment by dccsillag** [comment]
After the refactor in 3c36ba9d67a1ec11a824c98e044b9aaf315aa01e and e95fe7cbd8b7a4e8bff48b737146dda31870bf5c, and under normal load, we have:

```
       startup: 699.6
event                  time percent plot
BufEnter autocommand 298.34   42.65 ██████████████████████████
packer_compiled.lua   85.52   12.23 ███████▌
python3.vim           72.47   10.36 ██████▍
init.lua              48.41    6.92 ████▎
loading packages      25.07    3.58 ██▏
csillag.vim           21.15    3.02 █▉
filetype.vim          20.43    2.92 █▊
opening buffers       15.35    2.19 █▍
loading plugins       15.13    2.16 █▍
expanding arguments   11.58    1.65 █
indent_blankline.vim   8.73    1.25 ▊
syntax.vim             7.12    1.02 ▋
synload.vim            6.83    0.98 ▋
```

**Comment by dccsillag** [comment]
Without this plugin, here's what it looks like:

```
       startup: 171.6
event                  time percent plot
packer_compiled.lua   43.16   25.16 ██████████████████████████
init.lua              26.42   15.40 ███████████████▉
loading packages      16.64    9.70 ██████████
csillag.vim           11.62    6.78 ███████
filetype.vim          11.13    6.49 ██████▊
opening buffers        9.13    5.32 █████▌
loading plugins        8.96    5.23 █████▍
indent_blankline.vim   4.68    2.73 ██▉
syntax.vim             3.93    2.29 ██▍
synload.vim            3.75    2.19 ██▎
tcomment.vim           3.64    2.12 ██▎
csillag.vim            3.61    2.10 ██▏
```

**Comment by dccsillag** [comment]
Now back with the plugin, but with the `BufEnter` autocmd removed:

```
       startup: 329.7
event                  time percent plot
first screen update  137.31   41.64 ██████████████████████████
packer_compiled.lua   41.35   12.54 ███████▉
python3.vim           33.88   10.27 ██████▍
init.lua              24.08    7.30 ████▌
loading packages      15.48    4.69 ██▉
csillag.vim           10.77    3.27 ██
filetype.vim           9.94    3.02 █▉
loading plugins        8.56    2.60 █▋
opening buffers        7.48    2.27 █▍
indent_blankline.vim   4.03    1.22 ▊
syntax.vim             3.65    1.11 ▊
synload.vim            3.48    1.06 ▋
csillag.vim            3.35    1.02 ▋
rplugin.vim            2.87    0.87 ▌
BufEnter autocommand   2.84    0.86 ▌
tcomment.vim           2.72    0.82 ▌
rplugin.vim            2.71    0.82 ▌
```

It appears that there is _some_ initialization work done whenever the first thing regarding this plugin happens -- previously this first thing was BufEnter, now it's something in the first screen update.

**Comment by dccsillag** [comment]
I've removed magma from my config, and added an extremely simple plugin, based on the example on pynvim's documentation:

```python
import pynvim


@pynvim.plugin
class TestPlugin(object):
    def __init__(self, nvim):
        self.nvim = nvim

    @pynvim.function('TestFunction', sync=True)
    def testfunction(self, args):
        return 3

    @pynvim.command('TestCommand', nargs='*', range='')
    def testcommand(self, args, range):
        self.nvim.current.line = ('Command with args: {}, range: {}'
                                  .format(args, range))

    @pynvim.autocmd('BufEnter', sync=True)
    def on_bufenter(self):
        self.nvim.out_write("bufenter")
```

Startup time looks like this:

```
import pynvim


@pynvim.plugin
class TestPlugin(object):
    def __init__(self, nvim):
        self.nvim = nvim

    @pynvim.function('TestFunction', sync=True)
    def testfunction(self, args):
        return 3

    @pynvim.command('TestCommand', nargs='*', range='')
    def testcommand(self, args, range):
        self.nvim.current.line = ('Command with args: {}, range: {}'
                                  .format(args, range))

    @pynvim.autocmd('BufEnter', sync=True)
    def on_bufenter(self):
        self.nvim.out_write("bufenter")
```

Note the high amount of time spent on BufEnter; not quite magma-nvim levels, but still pretty high. And that's basically just Python remote plugin "initialization."

**Comment by dccsillag** [comment]
If I remove the `pynvim.autocmd` decorators from the plugin, the startup time becomes about the same as that of without the plugin. So I think the way is to only register the autocommands when `Magma._initialize` is called.

**Comment by dccsillag** [comment]
Solved in cfbfa4366f018fc3e87394747092134b9b5bc2f3.
